### PR TITLE
3536 get organisation by ods code integration tests

### PR DESCRIPTION
**Pull request by assoomro**

### CODE REVIEW

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature [hunk 1]
```diff
@@ -0,0 +1,87 @@
+﻿Feature: Display an Ods Organisation by its Ods Code
+  As an Authority User
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:2]: Tab this out correctly

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature [hunk 2]
```diff
@@ -0,0 +1,87 @@
+﻿Feature: Display an Ods Organisation by its Ods Code
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:1]: Display an Ods Organisation by its code

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+Background:
+	Given Ods Organisations exist
+		| Name               | OdsCode | PrimaryRoleId | Status    | AddrLn1 | AddrLn2         | AddrLn3         | AddrLn4     | Town     | County          | Postcode | Country |
+		| Ods Organisation 1 | B1G     | Buyer         | Active    | 12      | Brick Lane      | Central Area    | City Centre | Leeds    | West Yorkshire  | LS1 1AW  | England |
+		| Ods Organisation 2 | B2G     | Buyer         | Disbanded | 37      | Sand Brick Lane | Central Area    | City Centre | Bradford | West Yorkshire  | BD4 1FK  | England |
+		| Ods Organisation 3 | B3G     | AlsoBuyer     | Active    | 21      | Jump Street     | All over        | Suburb      | Metairie | Louisiana       | 504      | USA     |
+		| Ods Organisation 4 | N0G     | 153XP         | Active    | 15      | Sun Ave         | End of the Road | Suburb      | York     | North Yorkshire | YO11 4LO | England |
+	And Organisations exist
+		| Name           | OdsCode |
+		| Organisation 1 | Ods 1   |
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:15]: Do you not want to relate the OdsCode here to one of the Ods before?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		| Username             | Password        | Scope        |
+		| PostmanPat@email.com | An0therPa$$w0rd | Organisation |
+	When a GET request is made for an Ods organisation with code B1G
+	Then a response with status code 200 is returned
+	And the Ods Organisation is returned with the following values
+		| OrganisationName   | OdsCode | PrimaryRoleId | Line1 | Line2      | Line3        | Line4       | Town  | County         | Postcode | Country |
+		| Ods Organisation 1 | B1G     | Buyer         | 12    | Brick Lane | Central Area | City Centre | Leeds | West Yorkshire | LS1 1AW  | England |
+
+@3536
+Scenario: 2. Get the details of a single buyer organisation with different buyer role id
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:33]: Dont think you need this test, its doing the same thing as the test before

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	Given a user is logged in
+		| Username             | Password        | Scope        |
+		| PostmanPat@email.com | An0therPa$$w0rd | Organisation |
+	When a GET request is made for an Ods organisation with code B3G
+	Then a response with status code 200 is returned
+	And the Ods Organisation is returned with the following values
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:39]: Can then get rid of the B3G Ods code in the given

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+using WireMock.Admin.Mappings;
+using WireMock.Client;
+
+namespace NHSD.BuyingCatalogue.Identity.Api.IntegrationTests.Steps
+{
+    [Binding]
+    public sealed class OdsApiSteps
+    {
+        private readonly ScenarioContext _context;
+            private readonly Settings _settings;
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:17]: tabbing

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature (hunk 2 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:1]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature (hunk 1 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:2]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature (hunk 3 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:15]: I suppose I can 😄

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs (hunk 6 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:17]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        private async Task AddMapping(MappingModel model)
+        {
+            model.Priority = 10;
+            var api = RestClient.For<IWireMockAdminApi>(new Uri(_settings.OdsApiWireMockBaseUrl));
+            var result = await api.PostMappingAsync(model);
+            result.Status.Should().Be("Mapping added");
+            _context[ScenarioContextKeys.MappingAdded] = true;
+        }
+
+        public async Task ClearMappings()
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:74]: `public` methods should come before `private` methods, although I suspect this can probably be `internal` (which should come before `private` too).

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature (hunk 4 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:33]: I don't think its doing exactly the same thing. I agree - most of it is the same, however I included this test, to make sure our secondary 'primary role id' which identifies the buyer organisation works. In the config I've specified 'AlsoBuyer' primary role id to identify a buyer organisation and here I'm testing that it is correctly recognised as such. 

If you insist on removing this I'm happy to comply. Just let me know what do you think.

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            _settings = settings;
+        }
+
+        [Given(@"Ods API is down")]
+        public async Task CallToTheOdsApiWillFail()
+        {
+            await ClearMappings();
+            var model = CreateMappingModel("/ORD/2-0-0/organisations/*", 500);
+
+            await AddMapping(model);
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:31]: I'm not sure the `model` variable adds much value here. We could just do

```c#
await ClearMappings();
await AddMapping(CreateMappingModel("/ORD/2-0-0/organisations/*", 500));
```

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        public async Task SetUpGETEndpoint(string odsCode, string responseBody)
+        {
+            var model = CreateMappingModel($"/ORD/2-0-0/organisations/{odsCode}", 200, responseBody);
+
+            await AddMapping(model);
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:38]: I'm not sure the `model` variable adds much value here. We could just do:

```c#
await AddMapping(CreateMappingModel($"/ORD/2-0-0/organisations/{odsCode}", 200, responseBody));
```

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                            }
+                    },
+                    Methods = new[] { "GET" },
+                }
+            };
+        }
+
+        private async Task AddMapping(MappingModel model)
+        {
+            model.Priority = 10;
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:67]: What effect does this have?

File: src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs [hunk 11]
```diff
@@ -21,12 +21,12 @@ public OdsRepository(OdsSettings settings)
         public async Task<OdsOrganisation> GetBuyerOrganisationByOdsCodeAsync(string odsCode)
         {
             var response = await _settings.ApiBaseUrl
-                .AppendPathSegment("organisations")
+                .AppendPathSegments("ORD/2-0-0/organisations")
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs:24]: Could you explain why you decided to move the API version out of config?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                await _api.SetUpGETEndpoint(org.OdsCode, odsApiOrganisationAsJson);
+            }
+        }
+
+        [When(@"a GET request is made for an Ods organisation with code (.*)")]
+        public async Task WhenAGETRequestIsMadeForAnOdsOrganisationWithOdsCode(string odsCode)
+        {
+            using var client = new HttpClient();
+            client.SetBearerToken(_context.Get(ScenarioContextKeys.AccessToken, ""));
+            _response.Result = await client.GetAsync(new Uri($"{_organisationUrl}/{odsCode}"));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:51]: Would it simplify the API calls if we used Flurl?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        [Then(@"the Ods Organisation is returned with the following values")]
+        public async Task ThenTheOrganisationIsReturnedWithTheFollowingValues(Table table)
+        {
+            var expectedOrganisation = table.CreateSet<OrganisationTable>().FirstOrDefault();
+
+            JToken responseBody = await _response.ReadBodyAsJsonAsync();
+
+            var actualOrganisation = CreateOrganisationFromJson(responseBody);
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:61]: Would using Flurl simplify this?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        [Given(@"Ods API is down")]
+        public async Task CallToTheOdsApiWillFail()
+        {
+            await ClearMappings();
+            var model = CreateMappingModel("/ORD/2-0-0/organisations/*", 500);
+
+            await AddMapping(model);
+        }
+
+        public async Task SetUpGETEndpoint(string odsCode, string responseBody)
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:34]: Can this be `internal`?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs (hunk 7 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:74]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs (hunk 8 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:31]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs (hunk 9 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:38]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs (hunk 11 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs:24]: As I was doing the integration tests I came to question my decision of including this in the 'baseUrl' because of my association of baseUrls to bare urls of the servers.

Would you say there's any advantage of keeping it the way it used to be?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs (hunk 13 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:61]: With my limited knowledge I would lean towards saying no.

we could use its build-in serialization, however that would mean introducing yet another dto, (our 'Table' classes flatten out the 'Address' object so that they can be populated with specflow's tabular data with ease) and that might be slightly confusing to the reader.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs (hunk 11 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs:24]: The only real advantage that I can think of would be in the event of a minor (non-breaking) change to the ODS (say a bug fix or minor update that takes it to 2-0-1 or 2-1-0).

If we wanted to make use of the updated version we'd currently have to make a code change.  However, with the version in config it would just involve updating the setting/environment variable.

Admittedly, this may be something of an edge case though.

Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OdsRepository.cs:24]: I can see your point. and because the change is not that big, I'll revert to having the version included in the base url.

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs (hunk 10 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:67]: Absolutely none in this case, really good spot! 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs (hunk 14 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsApiSteps.cs:34]: it can, and will 😄

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs (hunk 12 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:51]: To compare

Now:

```    
      using var client = new HttpClient();
      client.SetBearerToken(_context.Get(ScenarioContextKeys.AccessToken, ""));
      _response.Result = await client.GetAsync(new Uri($"{_organisationUrl}/{odsCode}"));
```
With Flurl:
```
            _response.Result = await _organisationUrl
                .AppendPathSegment(odsCode)
                .WithOAuthBearerToken(_context.Get(ScenarioContextKeys.AccessToken, ""))
                .AllowAnyHttpStatus()
                .GetAsync();
```

Now, I quite like the fluent syntax, so they would look more pleasing to my eyes.

However, there are some cons
- we need to bring in another dependency 
- we'd probably need to refactor all the other operations using httpclient as to remain consistent (consistency rules)

Looking at the thing objectively, it will not be worth the refactor at the moment solely because I like it, however if more people agree that using flurl here is good, we can raise a task to refactor the tests (which shouldn't take that long).

TL;DR I'd like to leave it as is for now, discuss this with other devs and if they're in favour of refactoring, we refactor all the tests to use Flurl.

Thoughts? @bjss-chris

Comment (bjss-chris) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:51]: That sounds sensible to me @assoomro and thanks for preparing the example.

Comment (bjss-chris) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:51]: There's only one other tiny thing I'll mention here. I tend to prefer `string.Empty` over it's literal equivalent. The reason I prefer it is because it means I have no doubt about intent (i.e. the empty string isn't a mistake).

However, we don't currently have any standards covering this as far as I'm aware and as there is no functional difference I'll leave the decision up to you.

Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Steps/OdsOrganisationSteps.cs:51]: Good spot here, I agree that string.Empty would be better in this case. We can bring it up in today's code standards meeting along with the flurl proposition and hopefully raise TD tasks to clear this up (as it's used in all over the place).

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature [hunk 15]
```diff
@@ -84,4 +84,4 @@ Scenario: 8. Service Failure
 		| PostmanPat@email.com | An0therPa$$w0rd | Organisation |
 	And Ods API is down
 	When a GET request is made for an Ods organisation with code B1G
-	Then a response with status code 500 is returned
+	Then a response with status code 500 is returned
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature:87]: File should end with a newline to keep git happy.

Also, can you untabify the file? (Select All then Edit->Advanced->Untabify Selected Lines in VS.)

Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Ods Organisations GetByOdsCode 3536.feature]: Line endings seem to be a recurring theme of my PRs. I'll try watch out for this. 😅 
Thanks
